File: Daily/2131.longest-palindrome-by-concatenating-two-letter-words.js
/*
 * @lc app=leetcode id=2131 lang=javascript
 *
 * [2131] Longest Palindrome by Concatenating Two Letter Words
 */

// @lc code=start
/**
 * @param {string[]} words
 * @return {number}
 */
var longestPalindrome = function (words) {
    let length = 0;
    /** @type {Map<string, number>} */
    const wordToCount = new Map();
    for (const word of words) {
        wordToCount.set(word, (wordToCount.get(word) ?? 0) + 1);
    }

    const wordSet = new Set(words);

    for (const word of wordSet) {
        const wordCount = wordToCount.get(word) ?? 0;
        if (word[0] === word[1]) {
            const pairCount = Math.floor(wordCount / 2);
            length += pairCount * 4;
            wordToCount.set(word, wordCount - pairCount * 2);
        } else if (wordCount > 0) {
            const reversedWord = `${word[1]}${word[0]}`;
            const reversedWordCount = wordToCount.get(reversedWord) ?? 0;

            const reducedWordCount = Math.min(wordCount, reversedWordCount);
            length += 4 * reducedWordCount;
            wordToCount.set(word, wordCount - reducedWordCount);
            wordToCount.set(reversedWord, reversedWordCount - reducedWordCount);
        }
        if (wordToCount.get(word) === 0) {
            wordToCount.delete(word);
        }
    }

    for (const [word, count] of wordToCount) {
        if (word[0] === word[1] && count > 0) {
            length += 2;
            break;
        }
    }

    return length;
};
// @lc code=end

File: src/2025/2131.longest-palindrome-by-concatenating-two-letter-words.cpp
/*
 * @lc app=leetcode id=2131 lang=cpp
 *
 * [2131] Longest Palindrome by Concatenating Two Letter Words
 */

#include <algorithm>
#include <string>
#include <unordered_map>
#include <vector>

// @lc code=start
class Solution {
 public:
  int longestPalindrome(const std::vector<std::string>& words) {
    std::unordered_map<std::string, int> word_to_counts;
    for (const std::string& word : words) {
      word_to_counts[word]++;
    }

    int palindrome_length = 0;
    bool has_single_symmetric_word = false;
    for (const auto& [word, count] : word_to_counts) {
      if (word[0] > word[1]) {
        continue;
      }

      if (word[0] == word[1]) {
        has_single_symmetric_word = has_single_symmetric_word || count % 2;
        palindrome_length += 4 * (count / 2);
        continue;
      }
      const auto symmetric_word = std::string(word.crbegin(), word.crend());
      const int symmetric_word_count = word_to_counts.contains(symmetric_word)
                                           ? word_to_counts.at(symmetric_word)
                                           : 0;
      palindrome_length += 4 * std::min(count, symmetric_word_count);
    }
    if (has_single_symmetric_word) {
      palindrome_length += 2;
    }
    return palindrome_length;
  }
};
// @lc code=end
